templates/adding_a_new_model/open_model_proposals/ADD_GLM.md

```diff
@@ -0,0 +1,1129 @@
+🤗 Transformers implementation and the original implementation. Next, you
+should make sure that the output values are identical as well. This one
+of the most difficult parts of adding a new model. Common mistakes why
+the outputs are not identical are:
+
+-   Some layers were not added, *i.e.* an activation layer
+    was not added, or the residual connection was forgotten
+-   The word embedding matrix was not tied
+-   The wrong positional embeddings are used because the original
+    implementation uses on offset
+-   Dropout is applied during the forward pass. To fix this make sure
+    `model.training is False` and that no dropout layer is
+    falsely activated during the forward pass, *i.e.* pass
+    `self.training` to [PyTorch's functional
+    dropout](https://pytorch.org/docs/stable/nn.functional.html?highlight=dropout#torch.nn.functional.dropout)
+
+The best way to fix the problem is usually to look at the forward pass
+of the original implementation and the 🤗 Transformers implementation
+side-by-side and check if there are any differences. Ideally, you should
+debug/print out intermediate outputs of both implementations of the
+forward pass to find the exact position in the network where the 🤗
+Transformers implementation shows a different output than the original
+implementation. First, make sure that the hard-coded `input_ids` in both
+scripts are identical. Next, verify that the outputs of the first
+transformation of the `input_ids` (usually the word embeddings) are
+identical. And then work your way up to the very last layer of the
+network. At some point, you will notice a difference between the two
+implementations, which should point you to the bug in the 🤗 Transformers
+implementation. From our experience, a simple and efficient way is to
+add many print statements in both the original implementation and 🤗
+Transformers implementation, at the same positions in the network
+respectively, and to successively remove print statements showing the
+same values for intermediate presentions.
+
+When you're confident that both implementations yield the same output,
+verifying the outputs with
+`torch.allclose(original_output, output, atol=1e-3)`, you're done with
+the most difficult part! Congratulations - the work left to be done
+should be a cakewalk 😊.
+
+**8. Adding all necessary model tests**
+
+At this point, you have successfully added a new model. However, it is
+very much possible that the model does not yet fully comply with the
+required design. To make sure, the implementation is fully compatible
+with 🤗 Transformers, all common tests should pass. The Cookiecutter
+should have automatically added a test file for your model, probably
+under the same `tests/test_modeling_glm.py`. Run this test
+file to verify that all common tests pass:
+
+```python
+pytest tests/test_modeling_glm.py
+```
+
+[TODO FILL: Here the mentor should add very specific information on what tests are likely to fail after having implemented the model
```

Comment:
This is not super important, but would also be nice to be filled out, *e.g.* If GLM has a functionality that will probably make some common tests fail this should be noted here (e.g. BigBird has a very specific attention layer which made it clear that the general `test_attentions_output(...)` would fail).